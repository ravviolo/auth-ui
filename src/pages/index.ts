export * from './LoginPage/LoginPage';
export * from './SignUpPage/SignUpPage';
