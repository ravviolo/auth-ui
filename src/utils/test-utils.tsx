import { PreloadedState } from '@reduxjs/toolkit';
import { render, RenderOptions } from '@testing-library/react';
import React, { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import {
  create,
  ReactTestRenderer,
  TestRendererOptions,
} from 'react-test-renderer';

import { AppStore, RootState, setupStore } from 'store';

interface StoreOptions {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
  }: StoreOptions = {},
  renderOptions: RenderOptions = {}
) => {
  const Wrapper = ({ children }: PropsWithChildren<{}>): JSX.Element => {
    return (
      <BrowserRouter>
        <Provider store={store}>{children}</Provider>
      </BrowserRouter>
    );
  };

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};

export const createWithProviders = (
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
  }: StoreOptions = {},
  testRendererOptions?: TestRendererOptions
): ReactTestRenderer =>
  create(
    <BrowserRouter>
      <Provider store={store}>{ui}</Provider>
    </BrowserRouter>,
    testRendererOptions
  );
