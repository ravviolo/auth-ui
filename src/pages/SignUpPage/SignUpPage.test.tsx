import { screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { SignUpPage } from 'pages';
import { RootState } from 'store';
import { createWithProviders, renderWithProviders } from 'utils/test-utils';

describe('SignUpPage', () => {
  const testEmail = 'test@test.com';

  it("initially should have no user's data and hold user status as 'idle'", () => {
    const { store } = renderWithProviders(<SignUpPage />);

    expect(store.getState().user).toEqual({ userData: null, status: 'idle' });
  });

  it("after sign up should save user's data, change user's status to 'complete'", async () => {
    const { store } = renderWithProviders(<SignUpPage />);

    const template = screen.getByTestId('signup-template-test-id');
    const emailInputField = within(template).getByTestId(
      'input-field-email-test-id'
    );
    const passwordInputField = within(template).getByTestId(
      'input-field-password-test-id'
    );
    const signUpBtn = within(template).getByTestId('btn-submit-signup-test-id');

    await userEvent.type(emailInputField, testEmail);
    await userEvent.type(passwordInputField, 'password');
    await userEvent.click(signUpBtn);

    const userStore = store.getState().user;

    expect(userStore.status).toBe('complete');
    expect(userStore.userData?.firstname).toBe('Baby');
    expect(userStore.userData?.lastname).toBe('Yoda');
    expect(userStore.userData?.email).toBe(testEmail);
  });

  it("if user exist should update store with new user's data", async () => {
    const preloadedState: RootState = {
      user: {
        status: 'complete',
        userData: {
          avatarUrl: 'url',
          birthday: 'birthday',
          createdAt: 'createdAt',
          email: 'email',
          firstname: 'firstname',
          lastname: 'lastname',
          username: 'username',
          uuid: 'uuid',
        },
      },
    };

    const { store } = renderWithProviders(<SignUpPage />, { preloadedState });

    const template = screen.getByTestId('signup-template-test-id');
    const emailInputField = within(template).getByTestId(
      'input-field-email-test-id'
    );
    const passwordInputField = within(template).getByTestId(
      'input-field-password-test-id'
    );
    const signUpBtn = within(template).getByTestId('btn-submit-signup-test-id');

    await userEvent.type(emailInputField, testEmail);
    await userEvent.type(passwordInputField, 'password');
    await userEvent.click(signUpBtn);

    const userStore = store.getState().user;

    expect(userStore.userData?.firstname).toBe('Baby');
    expect(userStore.userData?.lastname).toBe('Yoda');
    expect(userStore.userData?.email).toBe(testEmail);
  });

  it('should match snapshot', () => {
    const tree = createWithProviders(<SignUpPage />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
