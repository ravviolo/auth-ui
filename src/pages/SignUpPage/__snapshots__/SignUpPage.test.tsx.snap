// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpPage should match snapshot 1`] = `
<main
  className="sc-dmRaPn dQxFcV"
  data-testid="signup-template-test-id"
>
  <div
    className="sc-jSMfEi bQrDqy sc-kgflAQ iSfrTS"
    data-testid="flex-signup-template-test-id"
    direction="column"
  >
    <header
      data-testid="auth-header-signup-template-test-id"
    >
      <h1
        className="sc-gKXOVf zAjwy"
        data-testid="auth-heading-test-id"
      >
        Sign Up
      </h1>
    </header>
    <form
      data-testid="signup-form-test-id"
      onSubmit={[Function]}
    >
      <div
        className="sc-jSMfEi bQrDqy sc-idiyUo wwKtw"
        data-testid="flex-signup-form-test-id"
        direction="column"
      >
        <div
          className="sc-jSMfEi bQrDqy sc-breuTD jjaaRC"
          data-testid="email-input-signup-form-test-id"
          direction="column"
        >
          <label
            className="sc-papXJ daxRFt"
            data-testid="input-label-email-test-id"
            htmlFor="email"
          >
            Email
          </label>
          <input
            className="sc-ftvSup dLJLzu"
            data-testid="input-field-email-test-id"
            id="email"
            name="email"
            type="email"
          />
        </div>
        <div
          className="sc-jSMfEi bQrDqy sc-ksZaOG cRa-dNN"
          data-testid="password-input-signup-form-test-id"
          direction="column"
        >
          <label
            className="sc-papXJ daxRFt"
            data-testid="input-label-password-test-id"
            htmlFor="password"
          >
            Password
          </label>
          <input
            className="sc-ftvSup dLJLzu"
            data-testid="input-field-password-test-id"
            id="password"
            name="password"
            type="password"
          />
        </div>
        <button
          className="sc-bczRLJ jXOWFG"
          data-testid="btn-submit-signup-test-id"
          type="submit"
        >
          Sign Up
        </button>
      </div>
    </form>
    <div
      className="sc-dkzDqf fBSlBa"
      data-testid="divider-signup-test-id"
    >
      <div
        className="sc-hKMtZM cJojKy"
      />
      <span
        className="sc-eCYdqJ fGmLST"
      >
        OR
      </span>
    </div>
    <div
      className="sc-jSMfEi iChCMC sc-evZas fBPmfS"
      data-testid="auth-socials-signup-template-test-id"
      direction="row"
    >
      <button
        className="sc-iBkjds cxbDaZ"
        color="#DB4437"
        data-testid="icon-btn-Google-test-id"
        onClick={[Function]}
        type="button"
      >
        G
      </button>
      <button
        className="sc-iBkjds bqhZzF"
        color="#4267B2"
        data-testid="icon-btn-Facebook-test-id"
        onClick={[Function]}
        type="button"
      >
        f
      </button>
      <button
        className="sc-iBkjds hAqsSt"
        color="#0077B5"
        data-testid="icon-btn-LinkedIn-test-id"
        onClick={[Function]}
        type="button"
      >
        in
      </button>
    </div>
    <footer
      className="sc-kDDrLX kktpMP"
      data-testid="auth-footer-signup-template-test-id"
    >
      <div
        className="sc-jSMfEi iChCMC sc-crXcEl bUmuaF"
        data-testid="flex-footer-test-id"
        direction="row"
      >
        <p>
          Already a user?
        </p>
        <button
          className="sc-jqUVSM ebpvV sc-iqcoie bPYbbt"
          data-testid="btn-footer-test-id"
          onClick={[Function]}
          type="button"
        >
          Login
        </button>
      </div>
    </footer>
  </div>
</main>
`;
